ably/hosts.go

```diff
@@ -0,0 +1,233 @@
+package ably
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ably/ably-go/ably/internal/ablyutil"
+)
+
+// RSC15
+type restHosts struct {
+	primaryFallbackHost string
+	opts                *clientOptions
+	cache               *hostCache
+	visitedHosts        []string
+	sync.Mutex
+}
+
+func newRestHosts(opts *clientOptions) *restHosts {
+	return &restHosts{
+		opts: opts,
+		cache: &hostCache{
+			duration: opts.fallbackRetryTimeout(),
+		},
+	}
+}
+
+func (restHosts *restHosts) getPrimaryHost() string {
+	return restHosts.opts.getPrimaryRestHost()
+}
+
+func (restHosts *restHosts) getFallbackHost() string {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+	hosts, _ := restHosts.opts.getFallbackHosts()
+	shuffledFallbackHosts := ablyutil.Shuffle(hosts)
```

Comment:
If we move this just above where it's used, we avoid shuffling unnecessarily if we return the primary host earlier.

Reply:
Moved